File: src/main/java/com/xpdustry/nucleus/concurrent/NucleusExecutors.java
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.concurrent;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ThreadFactory;

public final class NucleusExecutors {

    public static ThreadFactory newVirtualThreadFactory(final String name) {
        return Thread.ofVirtual().name("nucleus-" + name + "-", 0).factory();
    }

    public static ExecutorService newVirtualThreadPerTaskExecutor(final String name) {
        return Executors.newThreadPerTaskExecutor(newVirtualThreadFactory(name));
    }

    public static ScheduledExecutorService newSingleThreadScheduledExecutor(final String name) {
        return Executors.newSingleThreadScheduledExecutor(
                Thread.ofPlatform().name("nucleus-" + name).factory());
    }

    private NucleusExecutors() {}
}

File: src/main/kotlin/com/xpdustry/nucleus/concurrent/NucleusExecutors.kt
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.concurrent

import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ThreadFactory

object NucleusExecutors {
    fun newVirtualThreadFactory(name: String): ThreadFactory {
        return Thread.ofVirtual().name("nucleus-" + name + "-", 0).factory()
    }

    fun newVirtualThreadPerTaskExecutor(name: String): ExecutorService {
        return Executors.newThreadPerTaskExecutor(newVirtualThreadFactory(name))
    }

    fun newSingleThreadScheduledExecutor(name: String): ScheduledExecutorService {
        return Executors.newSingleThreadScheduledExecutor(Thread.ofPlatform().name("nucleus-" + name).factory())
    }
}
